README.rst
ClassyPie
=========

Python toolkit for easily classifying (grouping) data values.

Introduction
------------

Data classification algorithms are commonly used to group together data
values in order to simplify and highlight different aspects of the data
distribution.

ClassyPie implements and gives users easy access to many of the most
commonly used data classification algorithms. Can be used on any
sequence of values, including objects by specifying a key-function to
obtain the value.

The library was originally built as a convenient high-level wrapper
around Carston Farmer's "class\_intervals.py" script for QGIS, but has
since been improved and expanded with several new algorithms and new
convenience functions.

License
-------

This code is free to share, use, reuse, and modify according to the MIT
license, see license.txt

Credits
-------

-  Karim Bahgat
-  Carston Farmer

Changes
-------

0.1.0 (2017-04-23)
~~~~~~~~~~~~~~~~~~

-  First official version


CHANGES.txt

## Changes

### 0.1.0 (2017-04-23)

- First experimental version


testing/tester.py
import classypie as c
from random import randint
import pyagg

vals = [-500000+_**2 for _ in range(-100,1000)]

print "\nbreaks:"
for algo in ("equal","quantile","stdev","pretty","natural","headtail"):
  print algo, c.breaks(vals, algo)

print "\nsplit:"
for algo in ("equal","quantile","stdev","pretty","natural","headtail"):
  print algo
  for valrang,members in c.split(vals, algo):
    print valrang,len(members)

print "\nmembership:"
for valrang,members in c.membership(vals, [(-10000,100000),(-10000,0),(0,1000000)]):
  print valrang,len(members)


fdsf


grph = pyagg.graph.Histogram(vals, 5)
grph.draw(300,300,background="red").view()

grph = pyagg.graph.BarChart()
barlabels,bars = zip(*c.split(vals, "headtail"))
bars = [len(vals) for vals in bars]
barlabels = [str(rng) for rng in barlabels]
grph.add_category("bleh", barlabels=barlabels, bars=bars)
grph.draw(300,300,background="red").view()

print c.breaks(vals, "headtail")

import time
t = time.clock()
for each in c.split(vals, "headtail"): #[-77,-33,55]):
  print each
print time.clock()-t





testing/tester2.py

import classypie as cp

items = [v**2 for v in range(100)]

##cfier = cp.MultiClassifier(items)
##cfier.add_classification("equaltest",
##                       breaks="equal",
##                       fromval=1,
##                       toval=10)
##cfier.add_classification("naturaltest",
##                       breaks="natural",
##                       fromval=1,
##                       toval=10)
##cfier.add_classification("headtailtest",
##                       breaks="headtail",
##                       fromval=1,
##                       toval=10)
##
##for item,info in cfier:
##    print item,info


cfier = cp.Classifier(items,
                      breaks="pretty",
                      valuestops=[1,30],
                      #start=0,
                      #end=10000
                      )
print "hmm"
print cfier
for item,classval in cfier:
    print item,classval


testing/testrescale.py

import classypie as cp

vals = range(16)

for old,new in cp.rescale(vals, 0, 100):
    print old,new

for old,new in cp.rescale(vals, 100, 0):
    print old,new

for old,new in cp.rescale(vals, (0,100,0), (100,0,100)):
    print old,new

for old,new in cp.rescale(vals, (100,0,100), (0,100,0)):
    print old,new


upload.py
import pipy
 
packpath = "classypie"
pipy.define_upload(packpath,
                   author="Karim Bahgat",
                   author_email="karim.bahgat.norway@gmail.com",
                   license="MIT",
                   name="ClassyPie",
                   changes=["First official version"],
                   description="Python library for classifying/grouping data values.",
                   url="http://github.com/karimbahgat/ClassyPie",
                   keywords="data classification grouping algorithm scheme",
                   classifiers=["License :: OSI Approved",
                                "Programming Language :: Python",
                                "Development Status :: 4 - Beta",
                                "Intended Audience :: Developers",
                                "Intended Audience :: Science/Research",
                                'Intended Audience :: End Users/Desktop'],
                   )

pipy.generate_docs(packpath)
#pipy.upload_test(packpath)
#pipy.upload(packpath)



testclassipy.py

import classipy as cp

items = [v**2 for v in range(100)]

##cfier = cp.MultiClassifier(items)
##cfier.add_classification("equaltest",
##                       breaks="equal",
##                       fromval=1,
##                       toval=10)
##cfier.add_classification("naturaltest",
##                       breaks="natural",
##                       fromval=1,
##                       toval=10)
##cfier.add_classification("headtailtest",
##                       breaks="headtail",
##                       fromval=1,
##                       toval=10)
##
##for item,info in cfier:
##    print item,info


cfier = cp.Classifier(items,
                      breaks="pretty",
                      valuestops=[1,30],
                      #start=0,
                      #end=10000
                      )
print "hmm"
print cfier
for item,classval in cfier:
    print item,classval


classypie/main.py

from __future__ import division
from . import _breaks
import itertools
import math



class Classifier(object):
    
    def __init__(self, items, breaks, classvalues, key=None, **kwargs):
        self.items = items
        
        if isinstance(breaks, bytes):
            algo = breaks
            breaks = None
            
        else:
            algo = "custom"
            breaks = breaks
            
        self.algo = algo
        self.breaks = breaks
        self.classvalues = classvalues # the raw preinterpolated valuestops of the classvalues
        self.key = key
        self.kwargs = kwargs
        self.classvalues_interp = None # the final interpolated classvalues

        self.update()

    def __repr__(self):
        import pprint
        metadict = dict(algo=self.algo,
                        breaks=self.breaks,
                        classvalues_interp=self.classvalues_interp)
        return "Classifier object:\n" + pprint.pformat(metadict, indent=4)

    def update(self):
        # force update/calculate breaks and class values
        # mostly used internally, though can be used to recalculate
        if self.algo == "unique":
            self.classvalues_interp = self.classvalues

        elif self.algo == "proportional":
            self.classvalues_interp = [self.classvalues[0], self.classvalues[-1]]
            items,values = zip(*rescale(self.items,
                                       newmin=self.classvalues_interp[0],
                                       newmax=self.classvalues_interp[-1],
                                       key=self.key,
                                       **self.kwargs))
            itemvals = [self.key(item) for item in items]
            if self.classvalues_interp[0] < self.classvalues_interp[-1]:
                minval,maxval = min(itemvals), max(itemvals)
            else:
                minval,maxval = max(itemvals), min(itemvals)
            self.breaks = [minval,maxval]

        else:
            if self.algo != "custom":
                self.breaks = breaks(items=self.items,
                                    algorithm=self.algo,
                                    key=self.key,
                                    **self.kwargs)
            self.classvalues_interp = class_values(len(self.breaks)-1, # -1 because break values include edgevalues so will be one more in length
                                                   self.classvalues)

    def __iter__(self):
        # loop and yield items along with their classnum and classvalue
        
        if self.algo == "unique":
            if isinstance(self.classvalues_interp, dict):
                # only return specified uniqueval-classval pairs
                for uid,subitems in unique(self.items, key=self.key, **self.kwargs):
                    if uid in self.classvalues_interp:
                        classval = self.classvalues_interp[uid]
                        for item in subitems:
                            yield item,classval
            else:
                # eternally iterate over classvalues for each unique value
                def classvalgen ():
                    while True:
                        for classval in self.classvalues_interp:
                            yield classval
                classvalgen = classvalgen()
                for uid,subitems in unique(self.items, key=self.key, **self.kwargs):
                    classval = next(classvalgen)
                    for item in subitems:
                        yield item,classval

        elif self.algo == "proportional":
            for item,newval in rescale(self.items,
                                       newmin=self.classvalues_interp[0],
                                       newmax=self.classvalues_interp[-1],
                                       key=self.key,
                                       **self.kwargs):
                yield item,newval

        else:
            for valrange,subitems in split(self.items, self.breaks, key=self.key, **self.kwargs):
                midval = (valrange[0] + valrange[1]) / 2.0
                classinfo = self.find_class(midval)
                if classinfo is not None:
                    classnum,_ = classinfo
                    classval = self.classvalues_interp[classnum-1] # index is zero-based while find_class returns 1-based
                    for item in subitems:
                        yield item,classval

    def find_class(self, value):
        return find_class(value, self.breaks)


################################
            

def find_class(value, breaks):
    """
    Given a set of breakpoints, calculate which two breakpoints an input
    value is located between, returning the class number (1 as the first class)
    and the two enclosing breakpoint values. A value that is not between any of
    the breakpoints, ie larger or smaller than the break endpoints, is considered
    to be a miss and returns None. 
    """
    
    prevbrk = breaks[0]
    classnum = 1
    for nextbrk in breaks[1:]:
        if eval(bytes(prevbrk)) <= eval(bytes(value)) <= eval(bytes(nextbrk)):
            return classnum, (prevbrk,nextbrk)
        prevbrk = nextbrk
        classnum += 1
    else:
        # Value was not within the range of the break points
        return None

def class_values(classes, valuestops):
    """
    Return x number of class values linearly interpolated
    between a minimum and maximum value.

    - classes: Number of classes values to return. 
    - valuestops: can be either a single number or sequences of numbers
        where a classvalue will be interpolated for each sequence number,
        and so all sequences must be equally long. Thus, specifying the
        valuestops as rgb color tuples will create interpolated color gradients.
    """
    # special case
    if classes <= 1:
        #raise Exception("Number of classes must be higher than 1")
        return [valuestops[0]]
    if len(valuestops) < 2:
        raise Exception("There must be at least two items in valuestops for interpolating between")

    def _lerp(val, oldfrom, oldto, newfrom, newto):
        oldrange = oldto - oldfrom
        relval = (val - oldfrom) / float(oldrange)
        newrange = newto - newfrom
        newval = newfrom + newrange * relval
        return newval

    # determine appropriate interp func for either sequenes or single values
    
    if all(hasattr(valstop, "__iter__") for valstop in valuestops):
        _len = len(valuestops[0])
        if any(len(valstop) != _len for valstop in valuestops):
            raise Exception("If valuestops are sequences they must all have the same length")
        def _interpfunc(val):
            relindex = _lerp(classnum, 0, classes-1, 0, len(valuestops)-1)
            fromval = valuestops[int(math.floor(relindex))]
            toval = valuestops[int(math.ceil(relindex))]
            classval = [_lerp(relindex, int(relindex), int(relindex+1), ifromval, itoval)
                        for ifromval,itoval in zip(fromval,toval)]
            return classval
    else:
        def _interpfunc(classnum):
            relindex = _lerp(classnum, 0, classes-1, 0, len(valuestops)-1)
            fromval = valuestops[int(math.floor(relindex))]
            toval = valuestops[int(math.ceil(relindex))]
            classval = _lerp(relindex, int(relindex), int(relindex+1), fromval, toval)
            return classval
    
    # perform
    classvalues = []
    for classnum in range(classes):
        classval = _interpfunc(classnum)
        classvalues.append(classval)

    return classvalues

def breaks(items, algorithm, key=None, extrabreaks=None, exclude=None, minval=None, maxval=None, **kwargs):
    """
    Only get the break points, including the start and endpoint.
    """

    # ensure values are numeric
    def forcenumber(val):
        try:
            val = float(val)
            return val
        except:
            return None
    
    # sort by key
    if key:
        keywrap = lambda x: forcenumber(key(x))
    else:
        keywrap = forcenumber
        
    items = (item for item in items if keywrap(item) is not None)
    if exclude is not None:
        if not isinstance(exclude, (list,tuple)): exclude = [exclude]
        items = (item for item in items if keywrap(item) not in exclude)
    if minval is not None: items = (item for item in items if keywrap(item) >= minval)
    if maxval is not None: items = (item for item in items if keywrap(item) <= maxval)
    items = sorted(items, key=keywrap)
    values = [keywrap(item) for item in items]

    # get breaks
    func = _breaks.__dict__[algorithm]
    breaks = func(values, **kwargs)

    # insert extra breaks (list of single break values or pairs)
    if extrabreaks:
        for val in extrabreaks:
            oldbreaks = list(breaks)
            
            # insert single break value anywhere after first same or greater breakpoint
            # (remember that duplicate breakpoints will collect only that specific value)
            prevbrk = oldbreaks[0]
            i = 0
            for nextbrk in oldbreaks[1:]:
                if prevbrk <= val < nextbrk:
                    breaks.insert(i, val)
                    break
                else:
                    prevbrk = nextbrk
                i += 1
    
    return breaks

def split(items, breaks, key=None, exclude=None, minval=None, maxval=None, **kwargs):
    """
    Splits a list of items into n non-overlapping classes based on the
    specified algorithm. Values are either the items themselves or
    a value extracted from the item using the key function. 

    Arguments:

    - **items**: The list of items or values to classify.
    - **breaks**: List of custom break values, or the name of the algorithm to use.
        Valid names are:
        - histogram (alias for equal)
        - equal
        - quantile
        - pretty
        - stdev
        - natural
        - headtail
        - log (base-10, uses offset to handle 0s but not negative numbers)
    - **key** (optional): Function used to extract value from each item, defaults to None and treats item itself as the value. 
    - **classes** (optional): The number of classes to group the items into.
    - more...

    Returns:

    - All the input items reorganized into the groups/classes that they belong to,
    as a list of lists of items.
    """

    # ensure values are numeric
    def forcenumber(val):
        try:
            val = float(val)
            return val
        except:
            return None

    # sort and get key
    if key:
        keywrap = lambda x: forcenumber(key(x))
    else:
        keywrap = forcenumber
        
    items = (item for item in items if keywrap(item) is not None)
    if exclude is not None:
        if not isinstance(exclude, (list,tuple)): exclude = [exclude]
        items = (item for item in items if keywrap(item) not in exclude)
    if minval is not None: items = (item for item in items if keywrap(item) >= minval)
    if maxval is not None: items = (item for item in items if keywrap(item) <= maxval)
    items = sorted(items, key=keywrap)
    values = [keywrap(item) for item in items]

    # if not custom specified, get break values from algorithm name
    if isinstance(breaks, bytes):
        func = _breaks.__dict__[breaks]
        breaks = func(values, **kwargs)
    else:
        # custom specified breakpoints
        breaks = list(breaks)

    breaks_gen = (brk for brk in breaks)
    loopdict = dict()
    loopdict["prevbrk"] = next(breaks_gen)
    loopdict["nextbrk"] = next(breaks_gen)

    def find_class(item, loopdict=loopdict):
        val = keywrap(item)
        
##        while eval(bytes(val)) > eval(bytes(loopdict["nextbrk"])):
##            loopdict["prevbrk"] = loopdict["nextbrk"]
##            loopdict["nextbrk"] = next(breaks_gen)
##        if eval(bytes(loopdict["prevbrk"])) <= eval(bytes(val)) <= eval(bytes(loopdict["nextbrk"])):
##            return loopdict["prevbrk"],loopdict["nextbrk"]
        
##        if eval(bytes(val)) < loopdict["prevbrk"]:
##            # value lower than first class
##            return None
##        else:
##            while not (eval(bytes(loopdict["prevbrk"])) <= eval(bytes(val)) <= eval(bytes(loopdict["nextbrk"]))):
##                print eval(bytes(loopdict["prevbrk"])) , eval(bytes(val)) , eval(bytes(loopdict["nextbrk"]))
##                # increment breaks until value is between
##                loopdict["prevbrk"] = loopdict["nextbrk"]
##                loopdict["nextbrk"] = next(breaks_gen, None)
##                if loopdict["nextbrk"] == None:
##                    return None
##            # supposedly in between, so test and return class range
##            if eval(bytes(loopdict["prevbrk"])) <= eval(bytes(val)) < eval(bytes(loopdict["nextbrk"])):
##                return loopdict["prevbrk"],loopdict["nextbrk"]

        prevbrk = breaks[0]
        for i,nextbrk in enumerate(breaks[1:]):
            ###print val,i+1,len(breaks)-1
            if eval(bytes(val)) < eval(bytes(prevbrk)):
                return None
            elif eval(bytes(prevbrk)) <= eval(bytes(val)) < eval(bytes(nextbrk)):
                return prevbrk,nextbrk
            elif eval(bytes(prevbrk)) == eval(bytes(val)) == eval(bytes(nextbrk)):
                return prevbrk,nextbrk
            elif i+1==len(breaks)-1 and eval(bytes(val)) <= eval(bytes(nextbrk)):
                return prevbrk,nextbrk
            prevbrk = nextbrk
        else:
            return None

    for valrange,members in itertools.groupby(items, key=find_class):
        if valrange is not None:
            yield valrange, list(members)

def unique(items, key=None, only=None, exclude=None):
    """
    Bins all same values together, so all bins are unique.
    Only for ints or text values.

    If specified, only values listed in 'only' or not in 'exclude' will be returned. 
    """
    
    # sort and get key
    if key:
        pass
    else:
        key = lambda x: x

    if only:
        items = (item for item in items if key(item) in only)
    elif exclude:
        items = (item for item in items if key(item) not in exclude)

    items = sorted(items, key=key)

    for uniq,members in itertools.groupby(items, key=key):
        yield uniq, list(members)

    # maybe add remaining groups if none?
    # ... 

def membership(items, ranges, key=None):
    """
    Groups can be overlapping/nonexclusive and are based on custom ranges.
    """
    ###
    if not key:
        key = lambda x: x
    ###
    for _min,_max in ranges:
        valitems = ((key(item),item) for item in items)
        members = [item for val,item in valitems if val >= _min and val <= _max]
        yield (_min,_max), members

def rescale(items, newmin, newmax, key=None, only=None, exclude=None):
    """
    Iterates over all items, along with a new value for each.
    The new value is the item value rescaled to range from newmin to newmax. 
    """
    # sort and get key
    if key:
        pass
    else:
        key = lambda x: x

    pairs = ((item,key(item)) for item in items)

    if only:
        pairs = ((item,val) for item,val in pairs if val in only)
    elif exclude:
        pairs = ((item,val) for item,val in pairs if val not in exclude)

    items,values = zip(*pairs)

    oldmin, oldmax = min(values), max(values)

    def _lerp(val, oldfrom, oldto, newfrom, newto):
        oldrange = oldto - oldfrom
        relval = (val - oldfrom) / float(oldrange)
        newrange = newto - newfrom
        newval = newfrom + newrange * relval
        return newval

    # determine appropriate interp func for either sequenes or single values

    if hasattr(newmin, "__iter__") and hasattr(newmax, "__iter__"):
        if len(newmin) != len(newmax):
            raise Exception("If newmin/newmax are sequences they must both have the same length")
        def newval(val):
            return [_lerp(val, oldmin, oldmax, ifromval, itoval)
                        for ifromval,itoval in zip(newmin,newmax)]
    else:
        def newval(val):
            return _lerp(val, oldmin, oldmax, newmin, newmax)

    for item,val in zip(items, values):
        nv = newval(val)
        yield item, nv
    
    








classipy/_breaks.py
# original: Carson Farmer
# http://carsonfarmer.com/2010/09/playing-around-with-classification-algorithms-python-and-qgis/
# modified: Karim Bahgat, 2015

from __future__ import division
import math
import random



# Algorithms for value breakpoints

def histogram(values, **kwargs):
    """
    Alias for equal interval.
    """
    return equal(values, **kwargs)

def equal(values, classes=5, interval=None, anchor=None, clip=True, start=None, end=None):
    """
    Equal interval algorithm in Python
    
    Returns breaks based on dividing the range of 'values' into 'classes' parts,
    or by specifying the interval and/or anchorpoint to start the divisioning. 
    """
    #values = sorted(values) # maybe not needed as is already done main.py
    
    # auto detect value range from values if not custom specified
    if start == None:
        _min = min(values)
    else:
        _min = start
    if end == None:
        _max = max(values)
    else:
        _max = end
    
    if interval:
        # interval is custom specified
        if anchor is None:
            start = _min
        else:
            # find the starting point as the first interval that precedes the minimum
            start = anchor
            while start < _min-interval:
                # if lower, then increase up until right before min
                start += interval
            while start > _min:
                # if higher, then decrease until below min
                start -= interval

        # start creating breaks, ensuring to include min and max values
        k = start
        res = []
        while k < _max+interval:
            res.append(k)
            k += interval

        # clip first and last break to min and max
        if clip:
            res[0] = _min
            res[-1] = _max
        
    else:
        # calculate interval based on nr of classes
        unit = (_max - _min) / classes
        res = [_min + k*unit for k in range(classes+1)]

    return res

def log(values, classes=5):
    """
    Log classification algorithm.

    Returns break points at equal intervals of the log10 of input values.
    Handles 0s by adding 1 before log transforming. Negative values will raise Exception.
    """
    # log transform values
    logs = [math.log10(v+1) for v in values]

    # create breaks
    minval,maxval = min(logs),max(logs)
    interval = (maxval-minval)/float(classes)
    logbreaks = []
    cur = minval
    while cur <= maxval:
        logbreaks.append(cur)
        cur += interval

    # transform back
    breaks = [(10**log)-1 for log in logbreaks]
    
    return breaks
    
def quantile(values, classes=5):
    """
    Quantile algorithm in Python
    
    Returns values taken at regular intervals from the cumulative 
    distribution function (CDF) of 'values'.
    """

    #values = sorted(values) # maybe not needed as is already done main.py
    
    n = len(values)
    breaks = []
    for i in range(classes):
        q = i / float(classes)
        a = q * n
        aa = int(q * n)
        r = a - aa
        Xq = (1 - r) * values[aa] + r * values[aa+1]
        breaks.append(Xq)
    breaks.append(values[n-1])
    return breaks

def pretty(values, classes=5, start=None, end=None):
    """
    R's pretty algorithm implemented in Python
    Code based on R implementation from 'labeling' R package.
    
    Compute a sequence of about 'n+1' equally spaced 'round' values
    which cover the range of the values in 'values'.  The values are chosen
    so that they are 1, 2 or 5 times a power of 10.

    Returns a number of breaks not necessarily equal to 'classes' using 
    rpretty, but likely to be legible.

    Parameters:
        values : list of input values
        classes     : number of class intervals
    """

    # auto detect value range from values if not custom specified
    if start == None:
        dmin = min(values)
    else:
        dmin = start
    if end == None:
        dmax = max(values)
    else:
        dmax = end
        
    # begin
    n = classes
    min_n = int(n / 3) # Nonnegative integer giving the minimal number of intervals
    shrink_sml = 0.75  # Positive numeric by a which a default scale is shrunk 
                                         # in the case when range(x) is very small (usually 0).
    high_u_bias = 1.5  # Non-negative numeric, typically > 1. The interval unit 
                                         # is determined as {1,2,5,10} times b, a power of 10.
                                         # Larger high.u.bias values favor larger units
    u5_bias = 0.5 + 1.5 * high_u_bias
                                         # Non-negative numeric multiplier favoring 
                                         # factor 5 over 2.
    h = high_u_bias
    h5 = u5_bias
    ndiv = n

    dx = dmax - dmin
    if dx == 0 and dmax == 0:
        cell = 1.0
        i_small = True
        U = 1
    else:
        cell = max(abs(dmin), abs(dmax))
        if h5 >= 1.5 * h + 0.5:
            U = 1 + (1.0/(1 + h))
        else:
            U = 1 + (1.5 / (1 + h5))
        i_small = dx < (cell * U * max(1.0, ndiv) * 1e-07 * 3.0)

    if i_small:
        if cell > 10:
            cell = 9 + cell / 10
            cell = cell * shrink_sml
        if min_n > 1:
            cell = cell / min_n
    else:
        cell = dx
        if ndiv > 1:
            cell = cell / ndiv
    if cell < 20 * 1e-07:
        cell = 20 * 1e-07
    
    base = 10.0**math.floor(math.log10(cell))
    unit = base
    if (2 * base) - cell < h * (cell - unit):
        unit = 2.0 * base
        if (5 * base) - cell < h5 * (cell - unit):
            unit = 5.0 * base
            if (10 * base) - cell < h * (cell - unit):
                unit = 10.0 * base
    # Maybe used to correct for the epsilon here??
    ns = math.floor(dmin / unit + 1e-07)
    nu = math.ceil(dmax / unit - 1e-07)

    # Extend the range out beyond the data. Does this ever happen??
    while ns * unit > dmin + (1e-07 * unit):
        ns = ns-1
    while nu * unit < dmax - (1e-07 * unit):
        nu = nu+1
    # If we don't have quite enough labels, extend the range out to make more (these labels are beyond the data :( )
    k = math.floor(0.5 + nu-ns)
    if k < min_n:
        k = min_n - k
        if ns >= 0:
            nu = nu + k / 2
            ns = ns - k / 2 + k % 2
        else:
            ns = ns - k / 2
            nu = nu + k / 2 + k % 2
        ndiv = min_n
    else:
        ndiv = k
    graphmin = ns * unit
    graphmax = nu * unit
    count = int(math.ceil(graphmax - graphmin)/unit)
    res = [graphmin + k*unit for k in range(count+1)]
    if res[0] < dmin:
        res[0] = dmin
    if res[-1] > dmax:
        res[-1] = dmax
    return res

def stdev(values, classes=5):
    """
    Python implementation of the standard deviation class interval algorithm
    as implemented in the 'classInt' package available for 'R'.
    
    Returns breaks based on 'pretty' of the centred and scaled values of 'values',
    and may have a number of classes different from 'classes'.
    """

    sd2 = 0.0
    N = len(values)
    _min = min(values)
    _max = max(values)
    mean = sum(values) / N
    for i in values:
        sd = i - mean
        sd2 += sd * sd
    sd2 = math.sqrt(sd2 / N)
    res = pretty(values=None, classes=5, start=(_min-mean)/sd2, end=(_max-mean)/sd2)
    res2 = [(val*sd2)+mean for val in res]
    return res2

def natural(values, classes=5, maxsize=1000, samples=3):
    """
    Jenks Optimal (Natural Breaks) algorithm implemented in Python.
    The original Python code comes from here:
    http://danieljlewis.org/2010/06/07/jenks-natural-breaks-algorithm-in-python/
    and is based on a JAVA and Fortran code available here:
    https://stat.ethz.ch/pipermail/r-sig-geo/2006-March/000811.html
    
    Returns class breaks such that classes are internally homogeneous while 
    assuring heterogeneity among classes.

    For very large datasets (larger than maxsize), will calculate only on
    subsample to avoid exponential runtimes. Calculated multiple times (samples)
    and takes the average break values for better consistency. Lower and higher
    bounds are kept intact. 
    """

    #values = sorted(values) # maybe not needed as is already done main.py

    def getbreaks(values, classes):
        # the original algorithm by Carson Farmer
        mat1 = []
        for i in range(0,len(values)+1):
            temp = []
            for j in range(0,classes+1):
                temp.append(0)
            mat1.append(temp)
        mat2 = []
        for i in range(0,len(values)+1):
            temp = []
            for j in range(0,classes+1):
                temp.append(0)
            mat2.append(temp)
        for i in range(1,classes+1):
            mat1[1][i] = 1
            mat2[1][i] = 0
            for j in range(2,len(values)+1):
                mat2[j][i] = float('inf')
        v = 0.0
        for l in range(2,len(values)+1):
            s1 = 0.0
            s2 = 0.0
            w = 0.0
            for m in range(1,l+1):
                i3 = l - m + 1
                val = float(values[i3-1])
                s2 += val * val
                s1 += val
                w += 1
                v = s2 - (s1 * s1) / w
                i4 = i3 - 1
                if i4 != 0:
                    for j in range(2,classes+1):
                        if mat2[l][j] >= (v + mat2[i4][j - 1]):
                            mat1[l][j] = i3
                            mat2[l][j] = v + mat2[i4][j - 1]
            mat1[l][1] = 1
            mat2[l][1] = v
        k = len(values)
        kclass = []
        for i in range(0,classes+1):
            kclass.append(0)
        kclass[classes] = float(values[len(values) - 1])
        kclass[0] = float(values[0])
        countNum = classes
        while countNum >= 2:
            id = int((mat1[k][countNum]) - 2)
            kclass[countNum - 1] = values[id]
            k = int((mat1[k][countNum] - 1))
            countNum -= 1
        return kclass

    # Automatic sub sampling for large datasets
    # The idea of using random sampling for large datasets was in the original code. 
    # However, since these samples tend to produce different results,
    # ...to produce more stable results we might as well calculate the
    # ...breaks several times and using the sample means for the final break values.
    
    if len(values) > maxsize:
        allrandomsamples = []
        for _ in range(samples):
            randomsample = sorted(random.sample(values, maxsize))
            
            # include lower and higher bounds to ensure the whole range is considered
            randomsample[0] = values[0] 
            randomsample[-1] = values[-1]
            
            # get sample break
            tempbreaks = getbreaks(randomsample, classes)
            allrandomsamples.append(tempbreaks)
            
        # get average of all sampled break values
        jenksbreaks = [sum(allbreakvalues)/float(len(allbreakvalues))
                       for allbreakvalues in zip(*allrandomsamples)]
        
    else:
        jenksbreaks = getbreaks(values, classes)

    return jenksbreaks

def headtail(values, classes=5):
    """
    New head tails classification scheme,
    claimed to better highlight a few very
    large values than natural breaks.
    See: http://arxiv.org/ftp/arxiv/papers/1209/1209.2801.pdf
    """

    def _mean(values):
        return sum(values)/float(len(values))
    
    def _mbreak(values):
        m = _mean(values)
        head = [v for v in values if v >= m]
        tail = [v for v in values if v < m]
        return head,m,tail

    breaks = []
    head,m,tail = _mbreak(values)
    while len(tail) > len(head):
        breaks.append(m)
        if len(head) > 1:
            head,m,tail = _mbreak(head)
        else:
            break

    # add first and last endpoints
    breaks.insert(0, values[0])
    breaks.append(values[-1])
    
    # merge top breaks until under maxclasses
    if classes and len(breaks) > classes:
        pass
        
    return breaks

def auto(values, classes=5, **kwargs):
    pass




tester.py
import classypie as c
from random import randint

vals = [-500000+_**2 for _ in range(-100,1000)]

print("\nbreaks:")
for algo in ("equal","quantile","stdev","pretty","natural","headtail"):
  print(algo, c.breaks(vals, algo))

print("\nsplit:")
for algo in ("equal","quantile","stdev","pretty","natural","headtail"):
  print(algo)
  for valrang,members in c.split(vals, algo):
    print(valrang,len(members))

print("\nmembership:")
for valrang,members in c.membership(vals, [(-10000,100000),(-10000,0),(0,1000000)]):
  print(valrang,len(members))


fdsf


##import pyagg
##
##grph = pyagg.graph.Histogram(vals, 5)
##grph.draw(300,300,background="red").view()
##
##grph = pyagg.graph.BarChart()
##barlabels,bars = zip(*c.split(vals, "headtail"))
##bars = [len(vals) for vals in bars]
##barlabels = [str(rng) for rng in barlabels]
##grph.add_category("bleh", barlabels=barlabels, bars=bars)
##grph.draw(300,300,background="red").view()
##
##print c.breaks(vals, "headtail")
##
##import time
##t = time.clock()
##for each in c.split(vals, "headtail"): #[-77,-33,55]):
##  print each
##print time.clock()-t



